booster/library/Booster/Pattern/Rewrite.hs

```diff
@@ -340,41 +354,50 @@ applyRule pat@Pattern{ceilConditions} rule =
-                                (\v -> Var $ freshenVar v{variableName = stripMarker v.variableName} forbiddenVars)
-                                rule.existentials
-
-                    -- modify the substitution to include the existentials
-                    let substWithExistentials = subst `Map.union` existentialSubst
-
+                    -- partition ensured constrains into substitution and predicates
+                    let (newSubsitution, newConstraints) = partitionPredicates ensuredConditions
+
+                    -- merge existing substitution and newly acquired one, applying the latter to the former
+                    normalisedPatternSubst <-
+                        lift $
+                            normaliseSubstitution
+                                pat.substitution
+                                newSubsitution
+                    -- NOTE it is necessary to first apply the rule substitution and then the pattern/ensures substitution, but it is suboptimal to traverse the term twice.
+                    -- TODO a substitution composition operator
+                    let rewrittenTerm = substituteInTerm normalisedPatternSubst . substituteInTerm subst $ rule.rhs
+                        substitutedNewConstraints =
+                            Set.fromList $
+                                map
+                                    (coerce . substituteInTerm normalisedPatternSubst . substituteInTerm subst . coerce)
+                                    newConstraints
                     let rewritten =
                             Pattern
-                                (substituteInTerm substWithExistentials rule.rhs)
+                                rewrittenTerm
                                 -- adding new constraints that have not been trivially `Top`, substituting the Ex# variables
-                                ( pat.constraints
-                                    <> (Set.fromList $ map (coerce . substituteInTerm existentialSubst . coerce) newConstraints)
-                                )
+                                (pat.constraints <> substitutedNewConstraints)
+                                normalisedPatternSubst
                                 ceilConditions
                     withContext CtxSuccess $
                         withPatternContext rewritten $
                             return (rule, rewritten)
   where
+    -- Given known predicates, a known substitution and a newly acquired substitution (from the ensures clause):
+    -- - apply the new substitution to the old substitution
+    -- - merge with the new substitution items and return
+    normaliseSubstitution ::
+        Substitution -> Substitution -> RewriteT io Substitution
+    normaliseSubstitution oldSubst newSubst = do
+        let substitutedOldSubst = Map.map (substituteInTerm newSubst) oldSubst
+        pure (newSubst `Map.union` substitutedOldSubst) -- new bindings take priority
```

Comment:
THis effectively _composes_ substitutions IIUC.. could probably be used when constructing the result pattern, too.

Reply:
Renamed to `compose` and moved to `Booster.Pattern.Substitution`